mh/session.py

```diff
@@ -167,7 +167,9 @@ def get_layer_users(self, server_id, gate_id, city_id, first_index, count):
         return players[start:start+count]
 
     def leave_server(self):
-        DB.leave_server(self)
+        old_server = self.local_info["server_id"]
+        if old_server is not None:
```

Comment:
When does an error occur? In which circumstance this check is needed?

Reply:
This will be necessary for the implementation of a graceful connection-closed handler. When a user terminates their connection (such as by continuing to press B all the way back to Mode Select) from within the server select screen, they might or might not already be connected to a server.

Reply:
I still don't think that's the right approach to add complexity to this function to address the `recvReqShut` case. `session.layer_end` is most-likely a better candidate as `session.leave_server` was intended to be only called when needed (which is why there are asserts in the functions it calls).

I'd be fine with something like this, if needed:
```python
DB.leave_server(self)
self.local_info["server_id"] = None
```

Reply:
This part still seems superfluous to me:
```python
old_server = self.local_info["server_id"]
if old_server is not None:
```